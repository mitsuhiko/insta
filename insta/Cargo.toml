[package]
name = "insta"
version = "1.39.0"
license = "Apache-2.0"
authors = ["Armin Ronacher <armin.ronacher@active-4.com>"]
description = "A snapshot testing library for Rust"
edition = "2021"
rust-version = "1.60.0"
homepage = "https://insta.rs/"
repository = "https://github.com/mitsuhiko/insta"
keywords = ["snapshot", "testing", "jest", "approval"]
categories = ["development-tools::testing"]
readme = "README.md"
exclude = [
    "assets/*"
]

[package.metadata.docs.rs]
all-features = true
rustdoc-args = ["--cfg", "docsrs"]

[features]
default = ["colors"]

# when the redactions feature is enabled values can be redacted in serialized
# snapshots.
redactions = ["pest", "pest_derive", "serde"]

# Enables support for running filters on snapshot
filters = ["regex"]

# Glob support
glob = ["walkdir", "globset"]

# Color support
colors = ["console"]

# Serialization formats
# TODO: This could be cleaner by using "dep:csv" without renaming the dep, but
# this technique allows for a lower MSRV
csv = ["dep_csv", "serde"]
json = ["serde"]
ron = ["dep_ron", "serde"]
toml = ["dep_toml", "serde"]
yaml = ["serde"]

# internal feature exclusive to cargo-insta
_cargo_insta_internal = ["clap"]

[dependencies]
dep_csv = { package = "csv", version = "1.1.6", optional = true }
console = { version = "0.15.4", optional = true, default-features = false }
pest = { version = "2.1.3", optional = true }
pest_derive = { version = "2.1.0", optional = true }
dep_ron = { package = "ron", version = "0.7.1", optional = true }
dep_toml = { package = "toml", version = "0.5.7", optional = true }
globset = { version = "0.4.6", optional = true }
walkdir = { version = "2.3.1", optional = true }
similar = { version = "2.1.0", features = ["inline"] }
regex = { version = "1.6.0", default-features = false, optional = true, features = ["std", "unicode"] }
serde = { version = "1.0.117", optional = true }
linked-hash-map = "0.5.6"
lazy_static = "1.4.0"
# Not yet supported in our MSRV of 1.60.0
# clap = { workspace=true, optional = true }
clap = {version = "4.1", features = ["derive", "env"], optional = true}

[dev-dependencies]
rustc_version = "0.4.0"
serde = { version = "1.0.117", features = ["derive"] }
similar-asserts = "1.4.2"
