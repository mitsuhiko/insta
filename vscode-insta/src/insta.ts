import * as cp from "child_process";
import { Uri } from "vscode";
import { Snapshot } from "./Snapshot";

async function getPendingSnapshotsFor(results: Snapshot[], root: Uri): Promise<void> {
  return new Promise((resolve, reject) => {
    let buffer = "";
    const child = cp.spawn(
      "cargo",
      ["insta", "pending-snapshots", "--as-json"],
      {
        cwd: root.fsPath,
      }
    );
    if (!child) {
      reject(new Error("could not spawn cargo-insta"));
      return;
    }
    child.stdout?.setEncoding("utf8");
    child.stdout.on("data", (data) => (buffer += data));
    child.on("close", (_exitCode) => {
      for (const line of buffer.split(/\n/g)) {
        try {
          results.push(new Snapshot(root, JSON.parse(line)))
        } catch (e) {
          console.error(e);
        }
      }
      resolve();
    });
  });
}

export async function getPendingSnapshots(roots: Uri[]): Promise<Snapshot[]> {
  const results: Snapshot[] = [];
  await Promise.all(roots.map((root) => getPendingSnapshotsFor(results, root)));
  return results;
}

export function processInlineSnapshot(
  snapshot: Snapshot,
  op: "accept" | "reject"
): Promise<boolean> {
  if (!snapshot.inlineInfo) {
    return Promise.resolve(false);
  }
  return new Promise((resolve, reject) => {
    const child = cp.spawn("cargo", ["insta", op, "--snapshot", snapshot.key], {
      cwd: snapshot.rootUri.fsPath,
    });
    if (!child) {
      reject(new Error("could not spawn cargo-insta"));
      return;
    }
    child.on("close", (exitCode) => {
      resolve(exitCode === 0);
    });
  });
}

export async function processAllSnapshots(
  roots: Uri[],
  op: "accept" | "reject"
): Promise<boolean> {
  const results = await Promise.all(roots.map((rootUri) => new Promise<boolean>((resolve, reject) => {
    const child = cp.spawn("cargo", ["insta", op], {
      cwd: rootUri.fsPath,
    });
    if (!child) {
      reject(new Error("could not spawn cargo-insta"));
      return;
    }
    child.on("close", (exitCode) => resolve(exitCode === 0));
  })));
  return results.every((x) => x);
}
