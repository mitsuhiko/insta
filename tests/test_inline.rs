#[cfg(feature = "csv")]
use insta::assert_csv_snapshot;
#[cfg(feature = "ron")]
use insta::assert_ron_snapshot;
#[cfg(feature = "toml")]
use insta::assert_toml_snapshot;
use insta::{assert_debug_snapshot, assert_json_snapshot, assert_snapshot, assert_yaml_snapshot};
use serde::Serialize;
use std::thread;

#[test]
fn test_simple() {
    assert_debug_snapshot!(vec![1, 2, 3, 4], @r###"
    [
        1,
        2,
        3,
        4,
    ]
    "###);
}

#[test]
fn test_single_line() {
    assert_snapshot!("Testing", @"Testing");
}

#[test]
fn test_unnamed_single_line() {
    assert_snapshot!("Testing");
    assert_snapshot!("Testing-2");
}

// We used to use the thread name for snapshot name detection.  This is unreliable
// so this test now basically does exactly the same as `test_unnamed_single_line`.
#[test]
fn test_unnamed_thread_single_line() {
    let builder = thread::Builder::new().name("foo::lol::something".into());

    let handler = builder
        .spawn(|| {
            assert_snapshot!("Testing-thread");
            assert_snapshot!("Testing-thread-2");
        })
        .unwrap();

    handler.join().unwrap();
}

#[test]
fn test_newline() {
    // https://github.com/mitsuhiko/insta/issues/39
    assert_snapshot!("\n", @"
");
}

#[cfg(feature = "csv")]
#[test]
fn test_csv_inline() {
    #[derive(Serialize)]
    pub struct Email(String);

    #[derive(Serialize)]
    pub struct User {
        id: u32,
        username: String,
        email: Email,
    }

    assert_csv_snapshot!(User {
        id: 1453,
        username: "mehmed-doe".into(),
        email: Email("mehmed@doe.invalid".into()),
    }, @r###"
    id,username,email
    1453,mehmed-doe,mehmed@doe.invalid
    "###);
}

#[cfg(feature = "csv")]
#[test]
fn test_csv_inline_multiple_values() {
    #[derive(Serialize)]
    pub struct Email(String);

    #[derive(Serialize)]
    pub struct User {
        id: u32,
        username: String,
        email: Email,
    }

    let user1 = User {
        id: 1453,
        username: "mehmed-doe".into(),
        email: Email("mehmed@doe.invalid".into()),
    };
    let user2 = User {
        id: 1455,
        username: "mehmed-doe-di".into(),
        email: Email("mehmed@doe-di.invalid".into()),
    };

    assert_csv_snapshot!(vec![user1, user2], @r###"
    id,username,email
    1453,mehmed-doe,mehmed@doe.invalid
    1455,mehmed-doe-di,mehmed@doe-di.invalid
    "###);
}

#[cfg(feature = "ron")]
#[test]
fn test_ron_inline() {
    #[derive(Serialize)]
    pub struct Email(String);

    #[derive(Serialize)]
    pub struct User {
        id: u32,
        username: String,
        email: Email,
    }

    assert_ron_snapshot!(User {
        id: 42,
        username: "peter-doe".into(),
        email: Email("peter@doe.invalid".into()),
    }, @r###"
    User(
      id: 42,
      username: "peter-doe",
      email: Email("peter@doe.invalid"),
    )
    "###);
}

#[cfg(feature = "toml")]
#[test]
fn test_toml_inline() {
    #[derive(Serialize)]
    pub struct Email(String);

    #[derive(Serialize)]
    pub struct User {
        id: u32,
        username: String,
        email: Email,
    }

    assert_toml_snapshot!(User {
        id: 42,
        username: "peter-doe".into(),
        email: Email("peter@doe.invalid".into()),
    }, @r###"
    id = 42
    username = 'peter-doe'
    email = 'peter@doe.invalid'
    "###);
}

#[test]
fn test_json_inline() {
    assert_json_snapshot!(vec!["foo", "bar"], @r###"
    [
      "foo",
      "bar"
    ]
    "###);
}

#[test]
fn test_yaml_inline() {
    #[derive(Serialize)]
    pub struct User {
        id: u32,
        username: String,
        email: String,
    }

    assert_yaml_snapshot!(User {
        id: 42,
        username: "peter-pan".into(),
        email: "peterpan@wonderland.invalid".into()
    }, @r###"
    ---
    id: 42
    username: peter-pan
    email: peterpan@wonderland.invalid
    "###);
}

#[cfg(feature = "redactions")]
#[test]
fn test_yaml_inline_redacted() {
    #[derive(Serialize)]
    pub struct User {
        id: u32,
        username: String,
        email: String,
    }

    assert_yaml_snapshot!(User {
        id: 42,
        username: "peter-pan".into(),
        email: "peterpan@wonderland.invalid".into()
    }, {
        ".id" => "[user-id]"
    }, @r###"
    ---
    id: "[user-id]"
    username: peter-pan
    email: peterpan@wonderland.invalid
    "###);
}

#[test]
fn test_non_basic_plane() {
    assert_snapshot!("a ğŸ˜€oeu", @"a ğŸ˜€oeu");
}

#[test]
fn test_multiline_with_empty_lines() {
    assert_snapshot!("# first\nsecond\n  third\n\n# alternative", @r###"
    # first
    second
      third

    # alternative
    "###);
}
