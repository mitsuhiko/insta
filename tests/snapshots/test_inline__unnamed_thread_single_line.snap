---
source: tests/test_inline.rs
assertion_line: 40
expression: "\"Testing-thread\""

---
Testing-thread
